### ISSUE TITLE
Improving gaming on the web

### ISSUE THREAD

**Issue by nondebug** [web-platform-tests/interop#786]
### Description

MDN: [Gamepad API](https://developer.mozilla.org/en-US/docs/Web/API/Gamepad_API), [Pointer Lock API](https://developer.mozilla.org/en-US/docs/Web/API/Pointer_Lock_API), [Screen Orientation API](https://developer.mozilla.org/en-US/docs/Web/API/ScreenOrientation)

Specifications: [Gamepad API](https://www.w3.org/TR/gamepad/), [Pointer Lock API](https://w3c.github.io/pointerlock/), [Screen Orientation API](https://w3c.github.io/screen-orientation/) 

Games need reliable input APIs. **Gamepad API** and **Pointer Lock API** support are critically important for gaming on the web, especially for cloud gaming. **Screen Orientation API**, especially orientation locking, is important for gaming on mobile devices where orientation changes can disrupt gameplay.

Tasks:
* Fix [Gamepad API failing tests](https://wpt.fyi/results/gamepad) (excluding [idlharness-extensions.https.window.html](https://wpt.fyi/results/gamepad/idlharness-extensions.https.window.html))
* Fix [Pointer Lock API failing tests](https://wpt.fyi/results/pointerlock)
* Update test environments for Chromium, Firefox, and Safari to enable [Screen Orientation API locking tests](https://wpt.fyi/results/screen-orientation) to pass on wpt.fyi. Currently these tests fail when run on hosts that do not support screen orientation locking.

### Specification

Gamepad API, Pointer Lock API, Screen Orientation API are all Working Draft stage standards-track specifications in WebApps WG

### Additional Signals

This proposal is primarily driven by feedback from web platform developers regarding a use case (gaming) which is currently too niche to rank highly on surveys. Cloud gaming operators are heavily impacted by breakage in these APIs and have highlighted the need for better testing and interoperability.

The scope for this proposal is limited to fixing the existing failing tests, but is intended as as first step toward addressing the larger issue of poor test coverage. Automated testing is limited as testing most of the API surface would require implementing new test APIs to simulate connected peripherals (gamepads) or system capabilities (pointer locking, screen orientation locking). These test automation APIs should be implemented and new tests should be written in order to continue making progress on this focus area in future Interop iterations. Specifying new test APIs requires standards work and is out of scope for Interop, but can be pursued in parallel with this proposal.

**Comment by rejhgadellaa** [comment]
I think it's worth mentioning the [Fullscreen API](https://developer.mozilla.org/en-US/docs/Web/API/Fullscreen_API) here. With regards to support on iOS Safari, MDN reports:

> Only available on iPad, not on iPhone. Shows an overlay button which can not be disabled. Swiping down exits fullscreen mode, making it unsuitable for some use cases like games.
> [MDN: Fullscreen API](https://developer.mozilla.org/en-US/docs/Web/API/Fullscreen_API#browser_compatibility)

Also see: https://webventures.rejh.nl/blog/2024/history-of-safari-show-stoppers/#anchor--2020-add-fullscreen-api-to-ios-display-fullscreen

**Comment by evanjacover** [comment]
As we explore cloud gaming at Jackbox Games, these features do become more critical. We are currently taking the approach of limiting the browsers we support based on some of these features, especially screen orientation.

**Comment by AshleyScirra** [comment]
Many of the mentioned features are useful for web games made in [Construct](https://www.construct.net) so we generally support this, but especially gamepad. In particular it would be useful to have good support for button and axis events (see https://github.com/w3c/gamepad/pull/152)

**Comment by shalanah** [comment]
Related Proposals:
- https://github.com/web-platform-tests/interop/issues/732
- https://github.com/web-platform-tests/interop/issues/738
- https://github.com/web-platform-tests/interop/issues/718
- https://github.com/web-platform-tests/interop/issues/712

**Comment by xingri** [comment]
We, NVIDIA Geforce Now cloud gaming, strongly agree with the current proposal and wish to participate in activities that improve the compatibility and interoperability of APIs for cloud gaming on the web.

As an example of cloud gaming's reliance on the Pointer Lock API, all cloud gaming providers experienced a surge in crash issues last June, as shown in this Reddit post [Reddit post](https://www.reddit.com/r/GeForceNOW/comments/1egb9my/all_tabs_just_close_out_while_playing/).
 [This tracking issue](https://issues.chromium.org/issues/356196900) explains [the root cause](https://issues.chromium.org/issues/334271898).  Beyond unexpected crashes, the Pointer Lock API also suffers from compatibility issues across different browsers, as described in [this bug report](https://bugs.webkit.org/show_bug.cgi?id=272136).

We also advocate for better coverage and cross-browser compatibility for Gamepad APIs, as outlined in [the issue](https://github.com/w3c/gamepad/issues/191).
These examples highlight the very real challenges faced by developers and users due to inconsistencies and shortcomings in key web gaming APIs. Addressing these issues within Interop 2025 is crucial for the continued growth and success of web gaming.

**Comment by marcoscaceres** [comment]
Not sure if it's a stretch, but adding Screen Wake Lock into the mix might be good too. Agree that Fullscreen is pretty fundamental to all these, but could be handled by https://github.com/web-platform-tests/interop/issues/732.

**Comment by AshleyScirra** [comment]
Also note following Chromium bugs involving Pointer Lock:

- pointerrawupdate does not fire when mouse buttons change with Pointer Lock - https://issues.chromium.org/issues/40817651
- Pointer lock unadjustedMovement breaks pointerrawupdate buttons - https://issues.chromium.org/issues/353249342

**Comment by gsnedders** [comment]
> * Update test environments for Chromium, Firefox, and Safari to enable [Screen Orientation API locking tests](https://wpt.fyi/results/screen-orientation) to pass on wpt.fyi. Currently these tests fail when run on hosts that do not support screen orientation locking.

At least in the case of macOS, I'm pretty sure there is no concept of any orientation lock, for example, thus there's not really any way to update the test environment to achieve that. I don't know what the status quo for GNOME/Wayland or Windows is? (And changing to iOS doesn't seem like "updating" the test environment — that's a totally different test environment.)

**Comment by wpt-interop** [comment]
Thank you for proposing Improving gaming on the web for inclusion in Interop 2025.

We are pleased to let you know that this proposal was accepted as part of the Gaming investigation area. You can follow the progress of this investigation area on the [Interop 2025 dashboard](https://wpt.fyi/interop-2025).

For an overview of our process, see [proposal selection](https://github.com/web-platform-tests/interop/blob/main/2025/selection-process.md). Thank you for contributing to Interop 2025!

_Posted on behalf of the Interop team._

**Comment by AshleyScirra** [comment]
It looks like the investigation area is for the "Gamepad API testing" only. This proposal was a bit broader including Pointer Lock and Screen Orientation.

**Comment by foolip** [comment]
I have now created https://github.com/web-platform-tests/interop-gamepad. Thanks to @nondebug for agreeing to lead the investigation!
